web/packages/teleterm/src/services/tshd/createClient.ts

```diff
@@ -856,6 +863,71 @@ export default function createClient(
         );
       });
     },
+    getUserPreferences(
+      params: api.GetUserPreferencesRequest.AsObject,
+      abortSignal?: types.TshAbortSignal
+    ): Promise<api.UserPreferences.AsObject> {
+      return withAbort(abortSignal, callRef => {
+        const req = new api.GetUserPreferencesRequest().setClusterUri(
+          params.clusterUri
+        );
+
+        return new Promise((resolve, reject) => {
+          callRef.current = tshd.getUserPreferences(req, (err, response) => {
+            if (err) {
+              reject(err);
+            } else {
+              const res = response.toObject();
+              resolve(res.userPreferences);
+            }
+          });
+        });
+      });
+    },
+    updateUserPreferences(
+      params: api.UpdateUserPreferencesRequest.AsObject,
+      abortSignal?: types.TshAbortSignal
+    ): Promise<void> {
```

Comment:
> The way `authClient.UpdateUserPreferences()` work is that it expects a full preferences message. If any field is missing, it will be reset to the default value. Since we work on a subset in tshd, I need to fetch the previous preferences first and modify only the changed fields. This may seem a bit ineffective, but as I mentioned above, the only other way is to work on full preferences messages.

Where do you plan to do this? In the client or in tshd?

Reply:
@avatus Are there plans to support optimistic locking in user preferences? #30416

We should definitely use it here if it's available.

Reply:
> @avatus Are there plans to support optimistic locking in user preferences? #30416
> 
> 
> 
> We should definitely use it here if it's available.

I have no idea what optimistic locking is but sounds neat! I'll have to look into it when I get back on Monday